arch/risc-v/src/mpfs/mpfs_start.c

```diff
@@ -169,6 +175,16 @@ void __mpfs_start(uint64_t mhartid)
 
   mpfs_boardinitialize();
 
+#ifdef CONFIG_ARCH_USE_S_MODE
+  /* Initialize the per CPU areas */
+
+  if (mhartid != 0)
+    {
+      riscv_percpu_init();
```

Comment:
Yes, only the boot hart should do this, but there is no logic for selecting one yet. Need to fix this later if/when SMP is supported.

Reply:
If so, the check should be if (mhartid == 0) to avoid the double initialization

Reply:
The check is correct, hart0 for us at least runs the bootloader. Also, hart0 does not have supervisor mode at all, or FPU, this is why we check for hart0 in many places.

Need to somehow decide which hart 1-4 does the boot, and use that number here. But that comes when / if SMP support is added.

Reply:
If so, let's remove riscv_percpu_init which is unnecessary but complicate the code.

Reply:
Initializing the module is still necessary for S-mode even when using 1 hart, because the memory for the percpu elements lives there. The percpu structure is used to initialize sscratch which contains hartid (which cannot be read in S-mode), as well as the IRQ stack (only 1 at the moment, but it is ready for SMP too).

This works and is absolutely necessary when #ifdef CONFIG_ARCH_USE_S_MODE is true.

Reply:
Let's talk riscv_percpu_init in #5985